ouroboros-consensus/test-consensus/Test/Consensus/BlockchainTime.hs

```diff
@@ -0,0 +1,175 @@
+    , testProperty "slotAfterTime"         prop_slotAfterTime
+    , testProperty "delayNextSlot"         prop_delayNextSlot
+    ]
+
+{-------------------------------------------------------------------------------
+  Roundtrip properties for time conversion functions
+-------------------------------------------------------------------------------}
+
+-- > fixedToUTC (fixedFromUTC t) ~= t
+prop_UTC_Fixed_UTC :: UTCTime -> Property
+prop_UTC_Fixed_UTC t =
+      counterexample ("t':  " ++ show t')
+    . counterexample ("t'': " ++ show t'')
+    $ realToFrac (diffUTCTime t t'') < maxError
+  where
+    t'  = fixedFromUTC t
+    t'' = fixedToUTC   t'
+
+    maxError :: Double
+    maxError = 0.001 -- millisecond resolution
+
+-- > fixedFromUTC (fixedToUTC t) == t
+prop_Fixed_UTC_Fixed :: FixedUTC -> Property
+prop_Fixed_UTC_Fixed t = fixedFromUTC (fixedToUTC t) === t
+
+-- > fixedToNominal (fixedFromNominal d) ~= d
+prop_Nominal_Fixed_Nominal :: NominalDiffTime -> Property
+prop_Nominal_Fixed_Nominal d =
+      counterexample ("t':  " ++ show d')
+    . counterexample ("t'': " ++ show d'')
+    $ realToFrac (d - d'') < maxError
+  where
+    d'  = fixedDiffFromNominal d
+    d'' = fixedDiffToNominal   d'
+
+    maxError :: Double
+    maxError = 0.001 -- millisecond resolution
+
+-- > fixedFromNominal (fixedToNominal t) == t
+prop_Fixed_Nominal_Fixed :: FixedDiffTime -> Property
+prop_Fixed_Nominal_Fixed t = fixedDiffFromNominal (fixedDiffToNominal t) === t
+
+{-------------------------------------------------------------------------------
+  Properties of the time utilities
+-------------------------------------------------------------------------------}
+
+-- > slotAtTime (startOfSlot slot)) == (slot, 0)
+prop_startOfSlot :: SlotLength -> SystemStart -> Slot -> Property
+prop_startOfSlot slotLen start slot =
+      counterexample ("slotStart: " ++ show slotStart)
+    $ slotAtTime slotLen start slotStart === (slot, 0)
+  where
+    slotStart = startOfSlot slotLen start slot
+
+-- > now - slotLen < startOfSlot (fst (slotAtTime now)) <= now
+-- > 0 <= snd (slotAtTime now) < slotLen
+prop_slotAtTime :: SlotLength -> SystemStart -> FixedDiffTime -> Property
+prop_slotAtTime slotLen start execTime =
+      counterexample ("now:         " ++ show now)
+    . counterexample ("currentSlot: " ++ show currentSlot)
+    . counterexample ("timeSpent:   " ++ show timeSpent)
+    . counterexample ("slotStart:   " ++ show slotStart)
+    $    addFixedUTC (negate slotLen') now < startOfSlot slotLen start currentSlot
+    .&&. slotStart <= now
+    .&&. 0 <= timeSpent
+    .&&. timeSpent < getSlotLength slotLen
+  where
+    now                      = addFixedUTC execTime (getSystemStart start)
+    (currentSlot, timeSpent) = slotAtTime slotLen start now
+    slotStart                = startOfSlot slotLen start currentSlot
+    slotLen'                 = getSlotLength slotLen
+
+-- > 0 < fst (timeUntilNextSlot now) <= slotLen
```

Comment:
I was wondering if the edge cases (close to 0 and `== slotLen`) were also being tested. Using `tabulate "Edge case:" [edgeCase]` and the code below, I got the following:

```haskell
    edgeCase
      | timeLeft < 3                      = "close to 0"
      | timeLeft == getSlotLength slotLen = "== slotLen"
      | otherwise                         = "in the middle"
```

    Edge case: (10000 in total):
    56.25% in the middle
    43.74% close to 0
     0.01% == slotLen

I had to run 10000 tests to get the 0.01%, but at least this "proves" it is correct.

Reply:
Nice, John would be so proud :D I'll add this.